File: src/main/java/bad/robot/radiate/teamcity/JsonBuildUnmarshaller.java
package bad.robot.radiate.teamcity;

import bad.robot.http.HttpResponse;
import bad.robot.radiate.Unmarshaller;
import com.google.gson.Gson;

class JsonBuildUnmarshaller implements Unmarshaller<HttpResponse, Build> {

    @Override
    public Build unmarshall(HttpResponse response) {
        RunningBuild build = new Gson().fromJson(new JsonResponse(response).body(), RunningBuild.class);
        if (build.getRunInformation() == null)
            return new Build(build.getId(), build.getNumber(), build.getHref(), build.getStatusString(), build.getStatusText(), build.getStartDate(), build.getFinishDate(), build.getBuildType());
        return build;
    }
}

File: src/main/scala/bad/robot/radiate/teamcity/JsonBuildUnmarshaller.scala
package bad.robot.radiate.teamcity

import argonaut.Argonaut._
import simplehttp.HttpResponse
import bad.robot.radiate.{ParseError, Unmarshaller, Error}
import scalaz.syntax.std.either._
import scalaz.\/

class JsonBuildUnmarshaller extends Unmarshaller[HttpResponse, Build] {
  def unmarshall(response: HttpResponse): Error \/ Build = {
    val json = new JsonResponse(response).body
    json.decodeEither[Build].disjunction.leftMap(ParseError)
  }
}